tests/h/services/group_update_test.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import pytest
+import mock
+
+from sqlalchemy.exc import SQLAlchemyError
+
+from h.exceptions import APIError, ConflictError
+from h.services.group_update import GroupUpdateService
+from h.services.group_update import group_update_factory
+
+
+class TestGroupUpdate(object):
+
+    def test_it_updates_valid_group_attrs(self, factories, svc):
+        group = factories.Group()
+
+        svc.update(group, {
+            'name': 'foobar',
+            'description': 'I am foobar',
+        })
+
+        assert group.name == 'foobar'
+        assert group.description == 'I am foobar'
+
+    def test_it_returns_updated_group_model(self, factories, svc):
+        group = factories.Group()
+
+        updated_group = svc.update(group, {'name': 'whatnot'})
+
+        assert updated_group == group
+
+    def test_it_does_not_protect_against_undefined_properties(self, factories, svc):
+        group = factories.Group()
+
+        updated_group = svc.update(group, {'some_random_field': 'whatever'})
+
+        # This won't be persisted in the DB, of course, but the model instance
+        # doesn't have a problem with it
+        assert updated_group.some_random_field == 'whatever'
+
+    def test_it_raises_ValueError_if_name_fails_model_validation(self, factories, svc, db_session):
+        group = factories.Group()
+
+        with pytest.raises(ValueError, match='name must be between'):
+            svc.update(group, {
+                'name': 'ye'
+            })
+
+    def test_it_raises_ValueError_if_authority_provided_id_fails_model_validation(self, factories, svc, db_session):
+        group = factories.Group()
+
+        with pytest.raises(ValueError, match='must only contain characters allowed in encoded URIs'):
+            svc.update(group, {
+                'authority_provided_id': '%%^&#*'
+            })
+
+    def test_it_raises_conflict_error_on_provided_id_uniqueness_violation(self, factories, svc, db_session):
+        factories.Group(authority_provided_id='foo', authority='foo.com')
+        group2 = factories.Group(authority_provided_id='bar', authority='foo.com')
+
+        with pytest.raises(ConflictError, match='authority_provided_id'):
+            svc.update(group2, {
+                'authority_provided_id': 'foo'
+            })
+
+    def test_it_raises_APIError_if_unexpected_sqlalchemy_Exception_raised(self, factories):
+        group = factories.Group()
+
+        # Mocking or messing with the `db_session` itself breaks teardown
```

Comment:
This test is a bit tortuous. Let me know if there's a better way to write it.

Reply:
I commented above that maybe catching unexpected SQLAlchemy errors should be removed. If so we could remove this test too. Otherwise a mock session probably is the easiest way to test this.

You can make the mock in one line:

```python
session = mock.MagicMock(flush=mock.MagicMock(side_effect=SQLAlchemyError("nuh uh")))
```